src/coreclr/ilasm/sha256.h

```diff
@@ -0,0 +1,115 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+//*****************************************************************************
+// sha256.h
+//
+
+//
+// contains implementation of sha256 hash algorithm
+//
+//*****************************************************************************
+#ifndef __sha256__h__
+#define __sha256__h__
+
+#ifdef _WIN32
+inline HRESULT Sha256Hash(BYTE* pSrc, DWORD srcSize, BYTE* pDst, DWORD dstSize)
+{
+    if (dstSize < 32)
+    {
+        return E_FAIL;
+    }
+
+    BCRYPT_ALG_HANDLE  algHandle  = NULL;
+    BCRYPT_HASH_HANDLE hashHandle = NULL;
+
+    DWORD hashLength   = 0;
+    DWORD resultLength = 0;
+
+    NTSTATUS status = BCryptOpenAlgorithmProvider(&algHandle, BCRYPT_SHA256_ALGORITHM, NULL, 0);
+
+    if (!NT_SUCCESS(status))
+    {
+        goto cleanup;
+    }
+
+    status = BCryptGetProperty(algHandle, BCRYPT_HASH_LENGTH, (PBYTE)&hashLength, sizeof(hashLength), &resultLength, 0);
+
+    if (!NT_SUCCESS(status))
+    {
+        goto cleanup;
+    }
+
+    assert(hashLength == 32);
+    status = BCryptCreateHash(algHandle, &hashHandle, NULL, 0, NULL, 0, 0);
+
+    if (!NT_SUCCESS(status))
+    {
+        goto cleanup;
+    }
+
+    status = BCryptHashData(hashHandle, pSrc, srcSize, 0);
+
+    if (!NT_SUCCESS(status))
+    {
+        goto cleanup;
+    }
+
+    status = BCryptFinishHash(hashHandle, pDst, hashLength, 0);
+
+cleanup:
+    if (hashHandle != NULL)
+    {
+         BCryptDestroyHash(hashHandle);
+    }
+
+    if (algHandle != NULL)
+    {
+        BCryptCloseAlgorithmProvider(algHandle, 0);
+    }
+
+    return status;
+}
+#elif defined(__APPLE__)
+#include <CommonCrypto/CommonCrypto.h>
+#include <CommonCrypto/CommonDigest.h>
+
+inline HRESULT Sha256Hash(BYTE* pSrc, DWORD srcSize, BYTE* pDst, DWORD dstSize)
+{
+    if (dstSize < CC_SHA256_DIGEST_LENGTH)
+    {
+        return E_FAIL;
+    }
+
+    CC_SHA256(pSrc, (CC_LONG)srcSize, pDst);
+    return S_OK;
+}
+#else
+extern "C" {
+    #include "openssl.h"
+    #include "pal_evp.h"
+}
+
+inline bool IsOpenSslAvailable()
+{
+    return CryptoNative_OpenSslAvailable() != 0;
+}
+
+inline HRESULT Sha256Hash(BYTE* pSrc, DWORD srcSize, BYTE* pDst, DWORD dstSize)
+{
+    if (!IsOpenSslAvailable() || CryptoNative_EnsureOpenSslInitialized() || (dstSize < 32))
```

Comment:
nit: perhaps a bit simpler this way:
```suggestion
inline HRESULT Sha256Hash(BYTE* pSrc, DWORD srcSize, BYTE* pDst, DWORD dstSize)
{
    if (CryptoNative_OpenSslAvailable() == 0 || CryptoNative_EnsureOpenSslInitialized() != 0 || (dstSize < 32))
```

Reply:
I agree it looks a bit redundant in its current form. The intent behind `IsOpenSslAvailable` is to allow ILAsm to disable features based on OpenSSL availability, without leaking the `CryptoNative` API surface to it. Including it in this PR is probably premature, though I will want something like it when ILAsm starts using these wrappers.

Reply:
I thought I saw a comment from Jan K that said that we should treat OpenSSL here as required as it is for the platform... which is to say "it's required for 'Linux'" (which includes OpenBSD/NetBSD/etc).

If you're taking it as required, you can drop the call to CryptoNative_OpenSslAvailable; CryptoNative_EnsureOpenSslInitialized will do the same dlopen that Available does, it just fails (via `assert`) if the library isn't found.

If you're taking it as optional, then using OpenSslAvailable is correct.

Reply:
I suppose it makes sense to delegate the responsibility of checking for OpenSSL availability to the caller, if we're going to provide `IsOpenSslAvailable` anyway.